### PR TITLE
Adding authentication and authorization to Andes [do not merge]

### PR DESCRIPTION
**Pull request by akalankapagoda**

### CODE REVIEW

File: modules/andes-core/broker/src/main/java/org/wso2/andes/amqp/AMQPAuthenticationManager.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import javax.security.auth.login.LoginException;
+import java.io.UnsupportedEncodingException;
+import java.security.Principal;
+
+/**
+ * Manages authentication of QPID objects against Andes.
+ */
+public class AMQPAuthenticationManager {
+
+    public static final byte CREDENTIALS_SEPARATOR = (byte) 0;
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/amqp/AMQPAuthenticationManager.java]: needs comments.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/amqp/AMQPAuthenticationManager.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    public static final String RESPONSE_CHARSET = "UTF8";
+
+    private static AndesAuthenticationManager andesAuthenticationManager =
+                                            AndesContext.getInstance().getAndesAuthenticationManager();
+
+    public static Subject authenticate(byte[] response) throws LoginException {
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/amqp/AMQPAuthenticationManager.java]: need comments.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesContext.java [hunk 3]
```diff
@@ -41,6 +44,9 @@
     private static AndesContext instance = new AndesContext();
     private MessageStore messageStore;
     private int deliveryTimeoutForMessage;
+    private RealmService realmService;
+    private AndesAuthenticationManager andesAuthenticationManager;
```
Comment (ramith) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesContext.java]: comments.

File: modules/andes-core/common/src/main/java/org/wso2/andes/security/AMQPSecurityUtils.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.io.UnsupportedEncodingException;
+
+/**
+ * Utility class for AMQP security functions.
+ */
+public class AMQPSecurityUtils {
+
+    public static final String RESPONSE_CHARSET = "UTF8";
+    private static byte SEPARATOR = 0;
```
Comment (ramith) [inline modules/andes-core/common/src/main/java/org/wso2/andes/security/AMQPSecurityUtils.java]: comments.

File: modules/andes-core/common/src/main/java/org/wso2/andes/security/AMQPSecurityUtils.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+    public static byte[] encodeCredentials(String authenticationID, String password) throws AMQProtocolException {
```
Comment (ramith) [inline modules/andes-core/common/src/main/java/org/wso2/andes/security/AMQPSecurityUtils.java]: comments.

File: modules/andes-core/common/src/main/java/org/wso2/andes/security/AMQPSecurityUtils.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        try {
+            byte[] passwordBytes = password.getBytes(RESPONSE_CHARSET);
+            byte[] authnid = authenticationID.getBytes(RESPONSE_CHARSET);
+            byte[] encodedCredentials = new byte[passwordBytes.length + authnid.length + 2];
+            int size = 0;
+
+            encodedCredentials[size++] = SEPARATOR;
```
Comment (ramith) [inline modules/andes-core/common/src/main/java/org/wso2/andes/security/AMQPSecurityUtils.java]: would it be worthwhile to explain what is happening here.. (use of size variable)?

Comment (akalankapagoda) [inline modules/andes-core/common/src/main/java/org/wso2/andes/security/AMQPSecurityUtils.java]: +1. These are extracted from the qpid code, will explain the logic.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/amqp/AMQPAuthenticationManager.java (hunk 1 above)
Comment (akalankapagoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/amqp/AMQPAuthenticationManager.java]: Will add comments.

**Comment by akalankapagoda** [comment]
C5 implementation is on hold.

**Comment by ramith** [comment]
@akalankapagoda there is a chance we will revive this up.

**Pull request closed without merging**